examples/chat2-reliable/reliability.go

```diff
@@ -0,0 +1,458 @@
+		c.addToIncomingBuffer(msg)
+	}
+}
+
+func (c *Chat) processBufferedMessages() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	processed := make(map[string]bool)
+	remainingBuffer := make([]*pb.Message, 0, len(c.incomingBuffer))
+
+	for _, msg := range c.incomingBuffer {
+		if processed[msg.MessageId] {
+			continue
+		}
+
+		missingDeps := c.checkCausalDependencies(msg)
+		if len(missingDeps) == 0 {
+			// Release the lock while processing the message
+			c.mutex.Unlock()
+			if msg.Content != "" {
+				c.ui.ChatMessage(int64(c.getLamportTimestamp()), msg.SenderId, msg.Content)
+			}
+
+			c.addToMessageHistory(msg)
+			c.mutex.Lock()
+		} else {
+			remainingBuffer = append(remainingBuffer, msg)
+		}
+	}
+
+	c.incomingBuffer = remainingBuffer
+}
+
+func (c *Chat) reviewAckStatus(msg *pb.Message) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Review causal history
+	for _, msgID := range msg.CausalHistory {
+		for i, outMsg := range c.outgoingBuffer {
+			if outMsg.Message.MessageId == msgID {
+				// acknowledged and remove from outgoing buffer
+				c.outgoingBuffer = append(c.outgoingBuffer[:i], c.outgoingBuffer[i+1:]...)
+				break
+			}
+		}
+	}
+
+	// Review bloom filter
+	if msg.BloomFilter != nil {
+		receivedFilter := bloom.NewWithEstimates(bloomFilterSize, bloomFilterFPRate)
+		err := receivedFilter.UnmarshalBinary(msg.BloomFilter)
+		if err == nil {
+			for i := 0; i < len(c.outgoingBuffer); i++ {
+				if receivedFilter.Test([]byte(c.outgoingBuffer[i].Message.MessageId)) {
+					// possibly acknowledged and remove it from the outgoing buffer
+					c.outgoingBuffer = append(c.outgoingBuffer[:i], c.outgoingBuffer[i+1:]...)
+					i--
+				}
+			}
+		}
+	}
+}
+
+func (c *Chat) requestMissingMessage(messageID string) {
+	for retry := 0; retry < maxRetries; retry++ {
+		err := c.doRequestMissingMessageFromPeers(messageID)
+		if err == nil {
+			return
+		}
+
+		c.ui.ErrorMessage(fmt.Errorf("failed to retrieve missing message (attempt %d): %w", retry+1, err))
+
+		// Exponential backoff
+		delay := retryBaseDelay * time.Duration(1<<uint(retry))
+		if delay > maxRetryDelay {
+			delay = maxRetryDelay
+		}
+		time.Sleep(delay)
+	}
+
+	c.ui.ErrorMessage(fmt.Errorf("failed to retrieve missing message after %d attempts: %s", maxRetries, messageID))
+}
+
+func (c *Chat) checkCausalDependencies(msg *pb.Message) []string {
+	var missingDeps []string
+	for _, depID := range msg.CausalHistory {
+		if !c.bloomFilter.Test(depID) {
+			missingDeps = append(missingDeps, depID)
+		}
+	}
+	return missingDeps
+}
+
+func (c *Chat) addToMessageHistory(msg *pb.Message) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.messageHistory = append(c.messageHistory, msg)
+	c.messageHistory = c.resolveConflicts(c.messageHistory)
+
+	if len(c.messageHistory) > maxMessageHistory {
+		c.messageHistory = c.messageHistory[len(c.messageHistory)-maxMessageHistory:]
+	}
+}
+
+func (c *Chat) resolveConflicts(messages []*pb.Message) []*pb.Message {
```

Comment:
Unsure why we have to reprocess the entire `messageHistory`? Why not simply change the `append` function to consider only the incoming new message, add it at its lamport timestamp and only resolve any conflict if there is a matching timestamp entry already? There may be some go-language idiosyncrasies at play here that I'm missing, though. :)